File: gcs/mavlink/command_service.py
"""High-level vehicle commands (arm/disarm, mode changes, takeoff, fly-to).

This is the *policy* layer over ``ICommandSink``: it knows that arming is a
``MAV_CMD_COMPONENT_ARM_DISARM``, that "LOITER" maps to a Copter custom_mode,
and that "Fly To Here" is a GUIDED position target — but it does not know how
bytes reach the vehicle. It depends on the sink through a provider callable so it
never holds a stale reference across reconnects, reads live vehicle state through
another provider, and reports every action (and every failure) through the same
notice sink the decoder uses — so user-initiated errors show up in the message
log just like vehicle STATUSTEXT, Mission-Planner style.
"""
from __future__ import annotations

from typing import Callable, Optional

from pymavlink import mavutil

from ..domain import flight_modes
from ..domain.telemetry import Severity, StatusText, TelemetrySnapshot, _now_ms
from ..interfaces.ports import ICommandSink

MAV_MODE_FLAG_CUSTOM_MODE_ENABLED = 1
ARM_MAGIC_FORCE = 21196  # param2 value that forces (dis)arm past safety checks

LinkProvider = Callable[[], Optional[ICommandSink]]
StateProvider = Callable[[], TelemetrySnapshot]
NoticeSink = Callable[[StatusText], None]


class CommandService:
    def __init__(self, link_provider: LinkProvider,
                 state_provider: StateProvider,
                 notify: NoticeSink):
        self._link = link_provider
        self._state = state_provider
        self._notify = notify

    # ── arming ──────────────────────────────────────────────────────────────
    def arm(self, force: bool = False) -> None:
        self._send_arm(True, force)

    def disarm(self, force: bool = False) -> None:
        self._send_arm(False, force)

    def _send_arm(self, arm: bool, force: bool) -> None:
        sink = self._require_link()
        if sink is None:
            return
        verb = "ARM" if arm else "DISARM"
        try:
            sink.command_long(
                mavutil.mavlink.MAV_CMD_COMPONENT_ARM_DISARM,
                1.0 if arm else 0.0,
                float(ARM_MAGIC_FORCE) if force else 0.0,
            )
            self._info(f"{verb} command sent" + (" (forced)" if force else ""))
        except Exception as exc:  # pragma: no cover - hardware/runtime errors
            self._error(f"{verb} failed: {exc}")

    # ── flight modes ─────────────────────────────────────────────────────────
    def set_mode_by_name(self, mode_name: str) -> None:
        sink = self._require_link()
        if sink is None:
            return
        self._switch_mode(sink, mode_name, announce=True)

    # ── takeoff (GUIDED) ──────────────────────────────────────────────────────
    def takeoff(self, altitude_m: float) -> None:
        sink = self._require_link()
        if sink is None:
            return
        if not self._state().mode.armed:
            self._error("Takeoff refused: vehicle is DISARMED — arm first")
            return
        self._ensure_guided(sink)
        try:
            sink.command_long(mavutil.mavlink.MAV_CMD_NAV_TAKEOFF,
                              0.0, 0.0, 0.0, 0.0, 0.0, 0.0, float(altitude_m))
            self._info(f"Takeoff to {altitude_m:.0f} m")
        except Exception as exc:  # pragma: no cover
            self._error(f"Takeoff failed: {exc}")

    # ── fly to here (GUIDED position target) ──────────────────────────────────
    def fly_to(self, lat: float, lon: float, alt_rel: float) -> None:
        sink = self._require_link()
        if sink is None:
            return
        if not self._state().mode.armed:
            self._notify(StatusText(Severity.WARNING,
                                    "Fly-to sent while DISARMED — vehicle will ignore it",
                                    _now_ms(), True))
        self._ensure_guided(sink)
        try:
            sink.set_position_target_global(lat, lon, alt_rel)
            self._info(f"Fly to {lat:.6f}, {lon:.6f} @ {alt_rel:.0f} m")
        except Exception as exc:  # pragma: no cover
            self._error(f"Fly-to failed: {exc}")

    # ── run an uploaded mission (AUTO) ────────────────────────────────────────
    def start_mission(self) -> None:
        sink = self._require_link()
        if sink is None:
            return
        if not self._state().mode.armed:
            self._notify(StatusText(Severity.WARNING,
                                    "Start mission sent while DISARMED — arm first",
                                    _now_ms(), True))
        self._switch_mode(sink, "AUTO", announce=True)
        try:
            sink.command_long(mavutil.mavlink.MAV_CMD_MISSION_START, 0.0, 0.0)
            self._info("Mission start (AUTO)")
        except Exception as exc:  # pragma: no cover
            self._error(f"Mission start failed: {exc}")

    # ── helpers ───────────────────────────────────────────────────────────────
    def _ensure_guided(self, sink: ICommandSink) -> None:
        """Switch to GUIDED if we aren't already — guided commands need it."""
        s = self._state()
        current = flight_modes.mode_name(s.mode.autopilot, s.mode.custom_mode)
        if current != "GUIDED":
            self._switch_mode(sink, "GUIDED", announce=False)
            self._info("Mode → GUIDED (required for guided command)")

    def _switch_mode(self, sink: ICommandSink, mode_name: str, announce: bool) -> None:
        autopilot = self._state().mode.autopilot
        custom_mode = flight_modes.mode_id(autopilot, mode_name)
        if custom_mode is None:
            self._error(f"Unknown mode '{mode_name}' for this autopilot")
            return
        try:
            sink.set_mode(MAV_MODE_FLAG_CUSTOM_MODE_ENABLED, custom_mode)
            if announce:
                self._info(f"Mode → {mode_name}")
        except Exception as exc:  # pragma: no cover
            self._error(f"Set mode {mode_name} failed: {exc}")

    def _require_link(self) -> Optional[ICommandSink]:
        sink = self._link()
        if sink is None:
            self._error("Not connected — command ignored")
        return sink

    def _info(self, text: str) -> None:
        self._notify(StatusText(Severity.NOTICE, text, _now_ms(), True))

    def _error(self, text: str) -> None:
        self._notify(StatusText(Severity.ERROR, text, _now_ms(), True))

File: cpp/src/mavlink/command_service.cpp
#include "mavlink/command_service.h"

#include "domain/flight_modes.h"
#include "mavlink/mav_message.h"

#include <QTimer>

namespace gcs::mavlink {

using domain::Severity;
using domain::StatusText;
using domain::nowMs;
namespace fm = domain::flight_modes;

namespace {
constexpr int kCustomModeEnabled = 1; // MAV_MODE_FLAG_CUSTOM_MODE_ENABLED
constexpr float kArmMagicForce = 21196.0f; // param2 buộc (dis)arm bỏ qua kiểm tra
// Chờ phương tiện vào GUIDED trước khi cất cánh. Nếu gửi TAKEOFF ngay sau lệnh
// đổi chế độ (trong cùng chu kỳ xả outbox), autopilot nhận TAKEOFF trước khi kịp
// vào GUIDED và xác nhận vị trí → trả về "Need position estimate". Mission Planner
// không dính lỗi này vì có độ trễ thao tác tay giữa đổi mode và takeoff.
constexpr int kGuidedPollMs = 250;       // nhịp kiểm tra heartbeat báo GUIDED
constexpr int64_t kGuidedWaitMs = 3000;  // tối đa chờ vào GUIDED rồi bỏ cuộc
}

CommandService::CommandService(LinkProvider link, StateProvider state, NoticeSink notify)
    : m_link(std::move(link)), m_state(std::move(state)), m_notify(std::move(notify))
{
}

void CommandService::arm(bool force) { sendArm(true, force); }
void CommandService::disarm(bool force) { sendArm(false, force); }

void CommandService::sendArm(bool arm, bool force)
{
    auto *sink = requireLink();
    if (!sink)
        return;
    const QString verb = arm ? QStringLiteral("ARM") : QStringLiteral("DISARM");
    try {
        sink->commandLong(MAV_CMD_COMPONENT_ARM_DISARM,
                          arm ? 1.0f : 0.0f,
                          force ? kArmMagicForce : 0.0f);
        info(verb + QStringLiteral(" đã gửi") + (force ? QStringLiteral(" (buộc)") : QString()));
    } catch (const std::exception &e) {
        error(verb + QStringLiteral(" thất bại: ") + QString::fromUtf8(e.what()));
    }
}

void CommandService::setModeByName(const QString &modeName)
{
    auto *sink = requireLink();
    if (!sink)
        return;
    switchMode(sink, modeName, true);
}

void CommandService::takeoff(double altitudeM)
{
    auto *sink = requireLink();
    if (!sink)
        return;
    if (!m_state().mode.armed) {
        error(QStringLiteral("Từ chối cất cánh: phương tiện chưa ARM — hãy arm trước"));
        return;
    }
    const auto s = m_state();
    const QString current = fm::modeName(s.mode.autopilot, s.mode.customMode);
    if (current == QLatin1String("GUIDED")) {
        sendTakeoff(altitudeM); // đã ở GUIDED → cất cánh ngay
        return;
    }
    // Chưa ở GUIDED: chuyển chế độ rồi CHỜ heartbeat xác nhận mới gửi TAKEOFF,
    // tránh lỗi "Need position estimate" do lệnh tới trước khi vào GUIDED.
    ensureGuided(sink);
    info(QStringLiteral("Chờ vào GUIDED trước khi cất cánh…"));
    waitForGuidedThenTakeoff(altitudeM, nowMs() + kGuidedWaitMs);
}

void CommandService::waitForGuidedThenTakeoff(double altitudeM, int64_t deadlineMs)
{
    QTimer::singleShot(kGuidedPollMs, [this, altitudeM, deadlineMs] {
        if (!m_link()) {
            error(QStringLiteral("Mất kết nối — hủy cất cánh"));
            return;
        }
        const auto s = m_state();
        if (!s.mode.armed) {
            error(QStringLiteral("Phương tiện đã disarm — hủy cất cánh"));
            return;
        }
        const QString current = fm::modeName(s.mode.autopilot, s.mode.customMode);
        if (current == QLatin1String("GUIDED")) {
            sendTakeoff(altitudeM);
            return;
        }
        if (nowMs() >= deadlineMs) {
            error(QStringLiteral("Không vào được GUIDED sau %1 s — hủy cất cánh. "
                                 "Kiểm tra GPS/EKF (Need position estimate).")
                      .arg(kGuidedWaitMs / 1000.0, 0, 'f', 0));
            return;
        }
        waitForGuidedThenTakeoff(altitudeM, deadlineMs); // thử lại
    });
}

void CommandService::sendTakeoff(double altitudeM)
{
    auto *sink = requireLink();
    if (!sink)
        return;
    try {
        sink->commandLong(MAV_CMD_NAV_TAKEOFF, 0, 0, 0, 0, 0, 0,
                          static_cast<float>(altitudeM));
        info(QStringLiteral("Cất cánh lên %1 m").arg(altitudeM, 0, 'f', 0));
    } catch (const std::exception &e) {
        error(QStringLiteral("Cất cánh thất bại: ") + QString::fromUtf8(e.what()));
    }
}

void CommandService::flyTo(double lat, double lon, double altRel)
{
    auto *sink = requireLink();
    if (!sink)
        return;
    if (!m_state().mode.armed)
        m_notify(StatusText(Severity::Warning,
            QStringLiteral("Gửi bay-đến khi chưa ARM — phương tiện sẽ bỏ qua"),
            nowMs(), true));
    ensureGuided(sink);
    try {
        sink->setPositionTargetGlobal(lat, lon, altRel);
        info(QStringLiteral("Bay đến %1, %2 @ %3 m")
                 .arg(lat, 0, 'f', 6).arg(lon, 0, 'f', 6).arg(altRel, 0, 'f', 0));
    } catch (const std::exception &e) {
        error(QStringLiteral("Bay-đến thất bại: ") + QString::fromUtf8(e.what()));
    }
}

void CommandService::startMission()
{
    auto *sink = requireLink();
    if (!sink)
        return;
    if (!m_state().mode.armed)
        m_notify(StatusText(Severity::Warning,
            QStringLiteral("Bắt đầu nhiệm vụ khi chưa ARM — hãy arm trước"),
            nowMs(), true));
    switchMode(sink, QStringLiteral("AUTO"), true);
    try {
        sink->commandLong(MAV_CMD_MISSION_START, 0, 0);
        info(QStringLiteral("Bắt đầu nhiệm vụ (AUTO)"));
    } catch (const std::exception &e) {
        error(QStringLiteral("Bắt đầu nhiệm vụ thất bại: ") + QString::fromUtf8(e.what()));
    }
}

void CommandService::ensureGuided(interfaces::ICommandSink *sink)
{
    const auto s = m_state();
    const QString current = fm::modeName(s.mode.autopilot, s.mode.customMode);
    if (current != QLatin1String("GUIDED")) {
        switchMode(sink, QStringLiteral("GUIDED"), false);
        info(QStringLiteral("Chế độ → GUIDED (bắt buộc cho lệnh guided)"));
    }
}

void CommandService::switchMode(interfaces::ICommandSink *sink, const QString &modeName, bool announce)
{
    const int autopilot = m_state().mode.autopilot;
    const auto customMode = fm::modeId(autopilot, modeName);
    if (!customMode) {
        error(QStringLiteral("Chế độ không xác định '%1' cho autopilot này").arg(modeName));
        return;
    }
    try {
        sink->setMode(kCustomModeEnabled, *customMode);
        if (announce)
            info(QStringLiteral("Chế độ → %1").arg(modeName));
    } catch (const std::exception &e) {
        error(QStringLiteral("Đặt chế độ %1 thất bại: %2").arg(modeName, QString::fromUtf8(e.what())));
    }
}

interfaces::ICommandSink *CommandService::requireLink()
{
    auto *sink = m_link();
    if (!sink)
        error(QStringLiteral("Chưa kết nối — bỏ qua lệnh"));
    return sink;
}

void CommandService::info(const QString &text)
{
    m_notify(StatusText(Severity::Notice, text, nowMs(), true));
}

void CommandService::error(const QString &text)
{
    m_notify(StatusText(Severity::Error, text, nowMs(), true));
}

} // namespace gcs::mavlink
